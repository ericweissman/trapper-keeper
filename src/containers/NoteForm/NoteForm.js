import React, { Component } from 'react'
import { connect } from 'react-redux'
import { postNote } from '../../thunks/postNote'
import { Redirect, Link } from 'react-router-dom'
import { NoteItem } from '../../components/NoteItem/NoteItem'
import { editNote } from '../../thunks/editNote'
import { deleteNote } from '../../thunks/deleteNote'
import PropTypes from 'prop-types';
let shortID = require('short-id');

export class NoteForm extends Component {
  constructor(props) {
    super(props)
    this.state = {
      note: {
        title: this.props.note.title,
        id: this.props.note.id,
        timestamp: this.props.note.timestamp,
      },
      items: this.props.items,
      isDeleted: false,
      redirect: false
    }
  }

  componentDidMount = () => {
    !this.props.isEdit && this.setState({
      items: [{ id: shortID.generate(), description: '', noteID: this.state.note.id, timestamp: Date.now(), isCompleted: false }]
    })
  }

  handleTitleChange = (e) => {
    const { value } = e.target
    const { id, timestamp } = this.state.note
    this.setState({ note: { title: value, id, timestamp } })
  }

  handleItemChange = (e) => {
    const { name, value } = e.target
    const { items } = this.state
    const newItems = items.map(item => {
      if (item.id === name) {
        item.description = value
      }
      return item
    })
    this.setState({ items: newItems })
  }

  toggleComplete = (id) => {
    const { items } = this.state
    const newItems = items.map(item => {
      if (item.id === id) {
        item.isCompleted = !item.isCompleted
      }
      return item
    })
    this.setState({ items: newItems })
  }

  handleAddItem = (e) => {
    e.preventDefault()
    const { items, note } = this.state;
    this.setState({ items: [...items, { id: shortID.generate(), description: '', noteID: note.id, timestamp: Date.now(), isCompleted: false }] })
  }

  handleSubmit = (e) => {
    e.preventDefault()
    const { id, title } = this.state.note;
    const { items } = this.state;
    const { isEdit, editNote, postNote } = this.props;
    const url = isEdit ? `http://localhost:3001/api/v1/notes/${id}` : 'http://localhost:3001/api/v1/notes'
    isEdit ? editNote(url, { id, title, items }) : postNote(url, { id, title, items })
    this.setState({ redirect: true })
  }

  handleDelete = () => {
    const { id } = this.state.note
    const url = `http://localhost:3001/api/v1/notes/${id}`
    this.props.deleteNote(url, id)
    this.setState({ redirect: true })
  }

  handleItemDelete = (id) => {
    let items = this.state.items.filter(item => item.id !== id)
    this.setState({ items })
  }

  render() {
    let { title } = this.state.note
    const { items } = this.state
    const { isEdit } = this.props

    return (
      <form>
        {
          this.state.redirect && <Redirect to='/' />
        }
        {
          isEdit ? <button onClick={this.handleDelete}>delete</button>
            : <Link to='/'><button>Go back</button></Link>
        }
        <input onChange={this.handleTitleChange} name='title' value={title}></input>
        {
          items.map((item) => {
            if (!item.isCompleted) {
              return <NoteItem item={item} handleItemChange={this.handleItemChange} handleItemDelete={this.handleItemDelete} key={item.id} toggleComplete={this.toggleComplete} />
            }
          })
        }
        <div onClick={this.handleAddItem}>Add Item</div>
        <section>
          {
            items.map((item) => {
              if (item.isCompleted) {
                return <NoteItem item={item} handleItemChange={this.handleItemChange} handleItemDelete={this.handleItemDelete} key={item.id} toggleComplete={this.toggleComplete} />              }
            })
          }
        </section>
        <button onClick={this.handleSubmit}>SAVE</button>
      </form>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({
  postNote: (url, note) => dispatch(postNote(url, note)),
  deleteNote: (url, id) => dispatch(deleteNote(url, id)),
  editNote: (url, note) => dispatch(editNote(url, note))
})

NoteForm.propTypes = {
  deleteNote: PropTypes.func,
  editNote: PropTypes.func,
  isEdit: PropTypes.bool,
  items: PropTypes.array,
  note: PropTypes.object,
  postNote: PropTypes.func,
}

NoteForm.defaultProps = {
  isEdit: false,
  items: [],
  note: {},
}

export default connect(null, mapDispatchToProps)(NoteForm)