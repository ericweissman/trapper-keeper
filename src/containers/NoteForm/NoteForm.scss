.form-container {
  background-color: rgba(0, 0, 0, 0.7);
  left: 0;
  top: 0;
  height: 100%;
  width: 100%;
  position: fixed;
} 

.form {
  background-color: $primary-light;
  border: 1px solid black;
  padding: 20px;
  margin-left: -162.5px;
  margin-top: -200px;
  left: 50%;
  top: 50%; 
  box-shadow: 4px 4px 12px black;
  height: 400px;
  width: 325px;
  position: fixed;
  z-index: 12;
}