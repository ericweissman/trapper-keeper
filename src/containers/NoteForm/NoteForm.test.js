import { NoteForm, mapDispatchToProps } from '../NoteForm/NoteForm'
import React from 'react'
import { shallow } from 'enzyme'

const mockNote = { title: '', id: 1, timestamp: 555 }
const mockProps = {
  postNote: jest.fn(),
  editNote: jest.fn(),
  deleteNote: jest.fn(),
  note: mockNote,
  items: [],
  isEdit: false
}
let shortID = require('short-id')

describe('noteForm', () => {
  let wrapper
  Date.now = jest.fn().mockImplementation(() => 5)
  shortID.generate = jest.fn().mockImplementation(() => 6)

  beforeEach(() => {
    wrapper = shallow(
      <NoteForm {...mockProps} />
    )
  })

  describe('initial state', () => {

    it('should match the initial snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should have initial state', () => {
      const expectedState = {
        note: {
          id: 1,
          timestamp: 555,
          title: ''
        },
        items: [],
        isDeleted: false,
        redirect: false
      }

      expect(wrapper.state()).toEqual(expectedState)
    })
  })

  describe('handleTitleChange', () => {

    it('should setState of note title as a default', () => {
      const mockEvent = {
        target: {
          value: ''
        }
      }
      const expected = {
        title: '',
        id: 1,
        timestamp: 555,
      }

      wrapper.instance().handleTitleChange(mockEvent)
      expect(wrapper.state('note')).toEqual(expected)
    })

    it('should setState of note title and a blank item', () => {
      const mockEvent = {
        target: {
          value: 'k'
        }
      }
      const expected = {
        note: {
          title: 'k',
          id: 1,
          timestamp: 555,
        },
        items: [{
          description: '',
          id: 6,
          isCompleted: false,
          noteID: 1,
          timestamp: 5
        }],
        isDeleted: false,
        redirect: false,
      }

      wrapper.instance().handleTitleChange(mockEvent)
      expect(wrapper.state()).toEqual(expected)
    })
  })

  describe('handleItemChange', () => {

    it('should setState with item description', () => {
      const mockEvent = {
        target: {
          value: 'New Item Description',
          name: 6
        }
      }
      const initialItem = {
        description: '',
        id: 6,
        timestamp: 5,
        isCompleted: false,
        noteID: 1
      }
      const expected = [{
        description: 'New Item Description',
        id: 6,
        timestamp: 5,
        isCompleted: false,
        noteID: 1
      }, { ...initialItem, description: '' }]
      wrapper.setState({items: [initialItem]})
      wrapper.instance().handleItemChange(mockEvent)
      expect(wrapper.state('items')).toEqual(expected)
    })

    it('should not change state if no id is matched', () => {
      const mockEvent = {
        target: {
          value: 'New Item Description',
          name: 99
        }
      }
      const initialItem = [{
        description: '',
        id: 6,
        timestamp: 5,
        isCompleted: false,
        noteID: 1
      }]
      const expected = [{
        description: '',
        id: 6,
        timestamp: 5,
        isCompleted: false,
        noteID: 1
      }]
      wrapper.setState({ items: initialItem })
      wrapper.instance().handleItemChange(mockEvent)
      expect(wrapper.state('items')).toEqual(expected)
    })

    it('should not add more than one new item field at a time', () => {
      const mockEvent = {
        target: {
          value: 'New Item Description',
          name: 99
        }
      }
      const newItem = {
        description: '',
        id: 6,
        timestamp: 5,
        isCompleted: false,
        noteID: 1
      }
      const initialItem = [{
        description: 'need to add new item',
        id: 6,
        timestamp: 5,
        isCompleted: false,
        noteID: 1
      }]
      const expected = [{
        description: 'need to add new item',
        id: 6,
        timestamp: 5,
        isCompleted: false,
        noteID: 1
      }, newItem]
      wrapper.setState({ items: initialItem })
      wrapper.instance().handleItemChange(mockEvent)
      expect(wrapper.state('items')).toEqual(expected)
    })
  })

  describe('toggleComplete', () => {
    it('should toggle item as complete', () => {
      const mockID = 6
      const initialItem = [{
        description: '',
        id: 6,
        timestamp: 5,
        isCompleted: false,
        noteID: 1
      }]
      const expected = [{
        description: '',
        id: 6,
        timestamp: 5,
        isCompleted: true,
        noteID: 1
      }]
      wrapper.setState({items: initialItem})
      wrapper.instance().toggleComplete(mockID)
      expect(wrapper.state('items')).toEqual(expected)
    })

    it('should not change state if no id is matched', () => {
      const mockID = 99
      const initialItem = [{
        description: '',
        id: 6,
        timestamp: 5,
        isCompleted: false,
        noteID: 1
      }]
      const expected = [{
        description: '',
        id: 6,
        timestamp: 5,
        isCompleted: false,
        noteID: 1
      }]
      wrapper.setState({ items: initialItem })
      wrapper.instance().toggleComplete(mockID)
      expect(wrapper.state('items')).toEqual(expected)
    })
  })

  describe('handleSubmit', () => {
    it('should setState of redirect to true and call postNote when isEdit is false', () => {
      const mockEvent = {
        preventDefault: jest.fn()
      }
      const expected = true

      wrapper.instance().handleSubmit(mockEvent)
      expect(wrapper.state('redirect')).toEqual(expected)
      expect(wrapper.instance().props.postNote).toHaveBeenCalled()
    })

    it('should setState of redirect to true and call postNote when isEdit is true', () => {
      const mockEvent = {
        preventDefault: jest.fn()
      }
      const expected = true
      wrapper.setProps({ isEdit: true })
      wrapper.instance().handleSubmit(mockEvent)
      expect(wrapper.state('redirect')).toEqual(expected)
      expect(wrapper.instance().props.editNote).toHaveBeenCalled()
    })
  })

  describe('handleDelete', () => {
    it('should setState of redirect to true and call deleteNote', () => {
      const mockNote = {
        title: 'New Title',
        id: 1,
        timestamp: 555
      }
      wrapper.setState({note: mockNote})
      
      wrapper.instance().handleDelete()
      expect(wrapper.state('redirect')).toEqual(true)
      expect(wrapper.instance().props.deleteNote).toHaveBeenCalled()

    })
  })

  describe('handleItemDelete', () => {
    it('should setState of redirect to true and call deleteNote', () => {
      const mockID = 1
      const mockItems = [
        {
          description: 'Item 1',
          id: 1,
          timestamp: 5,
          isCompleted: false,
          noteID: 1
        },
        {
          description: 'Item 2',
          id: 2,
          timestamp: 6,
          isCompleted: false,
          noteID: 1
        }
      ]
      const expected = [{
        description: 'Item 2',
        id: 2,
        timestamp: 6,
        isCompleted: false,
        noteID: 1
      }]
      wrapper.setState({ items: mockItems })
      wrapper.instance().handleItemDelete(mockID)
      expect(wrapper.state('items')).toEqual(expected)
    })
  })

  describe('mapStateToProps', () => {

    it('should call dispatch when postNote is called', () => {
      const mockDispatch = jest.fn()
      const mockURL = 'http://localhost:3001/api/v1/notes'

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.postNote(mockURL, mockNote)
      expect(mockDispatch).toHaveBeenCalled()
    })

    it('should call dispatch when deleteNote is called', () => {
      const mockDispatch = jest.fn()
      const mockURL = 'http://localhost:3001/api/v1/notes'

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.deleteNote(mockURL, mockNote)
      expect(mockDispatch).toHaveBeenCalled()
    })

    it('should call dispatch when editNote is called', () => {
      const mockDispatch = jest.fn()
      const mockURL = 'http://localhost:3001/api/v1/notes'

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.editNote(mockURL, mockNote)
      expect(mockDispatch).toHaveBeenCalled()
    })
  })

})