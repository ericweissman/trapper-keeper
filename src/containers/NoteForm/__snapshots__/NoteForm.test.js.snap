// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`noteForm initial state should match the initial snapshot 1`] = `
<div>
  <Connect(NoteArea) />
  <div
    className="form-container"
  >
    <form
      className="form"
    >
      <div
        className="form-back-del-btn-container"
      >
        <Link
          replace={false}
          to="/"
        >
          <button
            className="form-back-del-btn"
          >
            Back
          </button>
        </Link>
      </div>
      <input
        autoComplete="off"
        className="title-input"
        name="title"
        onChange={[Function]}
        placeholder="Add your title here!"
        value=""
      />
      <div
        className="items-area"
      >
        <section
          className="incomplete-section"
        />
        <section
          className="completed-section"
        />
      </div>
      <button
        className="form-back-del-btn form-save-btn"
        onClick={[Function]}
      >
        Save
      </button>
    </form>
  </div>
</div>
`;
